src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -530,6 +539,59 @@ int sd_dhcp_client_set_max_attempts(sd_dhcp_client *client, uint64_t max_attempt
         return 0;
 }
 
+static sd_dhcp_option* dhcp_option_free(sd_dhcp_option *i) {
+        if (!i)
+                return NULL;
+
+        free(i->data);
+        return mfree(i);
+}
+
+int sd_dhcp_option_new(uint8_t option, void *data, size_t length, sd_dhcp_option **ret) {
+        _cleanup_(sd_dhcp_option_unrefp) sd_dhcp_option *p = NULL;
+        _cleanup_free_ void *q = NULL;
+
+        assert(ret);
+
+        p = new(sd_dhcp_option, 1);
+        if (!p)
+                return -ENOMEM;
+
+        q = memdup(data, length);
+        if (!q)
+                return -ENOMEM;
+
+        *p = (sd_dhcp_option) {
+                .n_ref = 1,
+                .option = option,
+                .length = length,
+                .data = TAKE_PTR(q),
+        };
+
+        *ret = TAKE_PTR(p);
+        return 0;
+}
+
+DEFINE_TRIVIAL_REF_UNREF_FUNC(sd_dhcp_option, sd_dhcp_option, dhcp_option_free);
+
+int sd_dhcp_client_set_dhcp_option(sd_dhcp_client *client, sd_dhcp_option *v) {
+        int r;
+
+        assert_return(client, -EINVAL);
+        assert_return(v, -EINVAL);
+
+        r = ordered_hashmap_ensure_allocated(&client->options, NULL);
```

Comment:
Right now all constructors and destructors are passed explicitly. It would be nicer to use
```C
DEFINE_HASH_OPS_FULL(dhcp_option_hash_ops, char, NULL, NULL, sd_dhcp_option*, sd_dhcp_option_unref);
```